### PR TITLE
qemu_disk_img_commit: add sync operation on guest after create file.

### PR DESCRIPTION
**Pull request by xiagao**
add sync operation on guest after create file on it.

Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

ID: 1379522

### CODE REVIEW

File: qemu/tests/qemu_disk_img_commit.py [hunk 1]
```diff
@@ -44,7 +44,11 @@ def run(test, params, env):
     commit_test.start_vm(n_params)
 
     # save file md5sum before commit
-    md5 = commit_test.save_file(t_file)
+    guest_sync = False
+    drive_cache = commit_test.params.get("drive_cache")
+    if drive_cache == 'none' or drive_cache == 'writeback':
```
Comment (pingl2016) [inline qemu/tests/qemu_disk_img_commit.py]: After copy the file from host to guest, qemu may be killed by qmp command. The file will not be flushed to disk whatever the drive cache mode. So I think it is better to sync directly.

File: qemu/tests/qemu_disk_img.py [hunk 2]
```diff
@@ -96,9 +96,15 @@ def __md5sum(self, dst):
...
 
     @error.context_aware
-    def save_file(self, dst):
+    def save_file(self, dst, guest_sync=False):
         error.context("save file('%s') md5sum in guest" % dst, logging.info)
         self.__create_file(dst)
+        if guest_sync:
+            login_timeout = int(self.params.get("login_timeout", 360))
+            session = self.vm.wait_for_login(timeout=login_timeout)
+            error.context("sync guest data after create it", logging.info)
```
Comment (pingl2016) [inline qemu/tests/qemu_disk_img.py]: Better to write as error.context("sync guest data after copy it", logging.info).
The file is created on the host, then copied to guest.

File: qemu/tests/qemu_disk_img.py [hunk 3]
```diff
@@ -99,6 +99,11 @@ def __md5sum(self, dst):
     def save_file(self, dst):
         error.context("save file('%s') md5sum in guest" % dst, logging.info)
         self.__create_file(dst)
+        login_timeout = int(self.params.get("login_timeout", 360))
+        session = self.vm.wait_for_login(timeout=login_timeout)
```
Comment (CongLi) [inline qemu/tests/qemu_disk_img.py]: Need judge the vm at first before using it.
if not self.vm:
    *****

File: qemu/tests/qemu_disk_img.py [hunk 4]
```diff
@@ -99,6 +99,11 @@ def __md5sum(self, dst):
...
+        error.context("sync guest data after copy it", logging.info)
+        session.cmd('sync')
```
Comment (CongLi) [inline qemu/tests/qemu_disk_img.py]: Need handle the exception here if cmd execution failed.

File: qemu/tests/qemu_disk_img.py (hunk 3 above)
Comment (pingl2016) [inline qemu/tests/qemu_disk_img.py]: It is checked in self.__create_file(), so i think the step can be skipped.

File: qemu/tests/qemu_disk_img.py [hunk 5]
```diff
@@ -99,6 +99,11 @@ def __md5sum(self, dst):
     def save_file(self, dst):
         error.context("save file('%s') md5sum in guest" % dst, logging.info)
         self.__create_file(dst)
+        login_timeout = int(self.params.get("login_timeout", 360))
+        session = self.vm.wait_for_login(timeout=login_timeout)
+        error.context("sync guest data after copy it", logging.info)
```
Comment (CongLi) [inline qemu/tests/qemu_disk_img.py]: It's better to specify "it".
Or:  only "sync guest data".

File: qemu/tests/qemu_disk_img.py (hunk 3 above)
Comment (CongLi) [inline qemu/tests/qemu_disk_img.py]: ok, it's fine. thanks @pingl2016

File: qemu/tests/qemu_disk_img.py [hunk 6]
```diff
@@ -99,6 +99,15 @@ def __md5sum(self, dst):
...
         error.context("save file('%s') md5sum in guest" % dst, logging.info)
         self.__create_file(dst)
+        login_timeout = int(self.params.get("login_timeout", 360))
+        session = self.vm.wait_for_login(timeout=login_timeout)
+        error.context("sync guest data")
+        cmd = "sync"
+        status, output = session.cmd_status_output(cmd)
+        if status != 0:
+            logging.error("Execute '%s' with failures('%s') " % (cmd, output))
+            return None
```
Comment (pingl2016) [inline qemu/tests/qemu_disk_img.py]: When guest shutdown abnormally like quit by qmp after copy a file from host to guest, the file may not be flushed to disk which will make the test failed. So we need to do sync here . If the command failed to issue, we need to raise TestFail exception.
One simple way is use session.cmd('sync'). if the command failed, exception ShellCmdError  will be raised.
The other way is as below:
status, output = session.cmd_status_output(cmd)
if status != 0:
    raise error.TestFail("Execute '%s' with failures('%s') " % (cmd, output))

Comment (pingl2016) [inline qemu/tests/qemu_disk_img.py]: As every caller checked the return code of function save_file(), so ignore the former comment.

**Comment by xiagao** [comment]
@CongLi ,@pingl2016

Thanks for your comment. have updated.

**Comment by pingl2016** [comment]
It is ok for me now.

**Comment by CongLi** [comment]
Acked-by: Cong Li <coli@redhat.com>

**Comment by CongLi** [comment]
Thanks @xiagao @pingl2016 .

**Pull request merged**